### ISSUE TITLE
EVA possible without EVA suits

### ISSUE THREAD

**Issue by rnyoakum** [raceintospace/raceintospace#488]
_TLDR: There is an unlikely, yet plausible, situation where a mission will have a player perform an EVA even though the player has not purchased the EVA program._

Normally, the VIB/VAB requires the player to have all the necessary equipment to perform a mission before mission may be launched (with the special case of Docking Modules, which may be checked at launch). For EVA missions, this includes having purchased an EVA program. Consequently, the mission execution code assumes that all required mission hardware is valid and correctly assigned. This holds true - for the planned steps. It does not for emergency steps.

There are several missions - such as the MANNED EARTH ORBITAL LM TEST - where it is possible to have an emergency spacewalk in response to a crisis, normally prompted by a LM failure. Some of these, such as the Manned Earth Orbital LM Test, are not considered EVA missions because they do not require an EVA for successful execution. Consequently, they do not require the EVA suit program to be purchased before the mission may be scheduled.

If the player has not purchased the EVA suit program before launching an affected mission, and is forced to make an emergency EVA, the result is currently _undefined_. As implemented, I think the game will skip the relevant step as if the spacewalk was successful. That's the safer possibility, but still undesirable ("Just how long _can_ John Glenn hold his breath?"). However, as mission code gets modified that could change; this is an unaddressed error.

Since players will normally perform a spacewalk well in advance of a LM testing, this is unlikely to occur, but it looks possible.

**Comment by peyre** [comment]
Wow, that's a good point.  If that happens, the EVA step should, of course, automatically fail.  Programming that could be tricky, I imagine.

**Comment by rnyoakum** [comment]
It's a particularly annoying exception. It's even more complicated than just failing the step: the EVA step needs to not only fail, but **catastrophically** fail. Even using the standard failure system and forcing a specific EVA error has issues, because the error messages are designed on the assumption the astronauts/cosmonauts are dressed for the occasion.

There are a couple of ideas I have, but I don't like either of them; they're both evil kludges. So this issue may wait until I've nothing better to work on or until I come up with a clever solution.

**Comment by peyre** [comment]
I don't see why it couldn't wait for the following release.  It's pretty obscure.

I don't suppose you could simply have it roll against a 0% safety?

**Comment by rnyoakum** [comment]
> I don't suppose you could simply have it roll against a 0% safety?

You could - except some of the randomly chosen failure cases will translate into success (Example: "OVEREXERTION, CREWMAN IS SEVERELY EXHAUSTED, BUT MAKES IT BACK."). You'd have to force the right failure case, and at its worst an EVA failure only kills the EVA specialist - insufficient for a two-person LM crew trying to return to the capsule.

**Comment by peyre** [comment]
Ohhh....I feel like doing a Picard facepalm here.  Of course.

**Comment by hweimer** [comment]
I guess the proper way to handle this is to not try an EVA step at all, but promote the previous failure to type 31 (i.e., kill everyone on the LM). Can this be done when the branching to emergency EVA is about to occur?

**Comment by rnyoakum** [comment]
Probably so - you could add a very specific check to FailEval right after evaluating the error code, to see if it is
* branching to the error case
* the next step is an EVA step
* there is no EVA suit assigned to the Mission_EVA hardware slot

and, if so, return a recursive call to FailEval with type=31.

That solution leans on the EVA step always being the next step, otherwise it would get really messy - but I think it's a reliable assumption.

The specificity of the check, right in the middle of the generalized FailEval() function, makes it a Very Ugly Hack, for a very rare problem. It would be more elegant to deal with the problem of "a mission check where no valid hardware exists," yet there are faux missions steps used for flow control that have no hardware associated with them.

**Comment by peyre** [comment]
Or maybe it could be like the (Orbit) missions - on the turn that you assemble the mission, if EVA suits aren't available, it  would drop the EVA part of the mission.

It might also be good to add a warning at the end of your turn that you have no EVA suits program, if you have a Manned Orbital or any EVA mission planned for next turn.  I'll add that to the End-of-Turn Warnings issue.

**Comment by peyre** [comment]
Wait a minute, is this even a real issue?  Future Missions won't let you schedule an EVA mission if you haven't started an EVA program.  OTOH, a warning when your suits are below Max R&D might be a good idea.

![NoEVA](https://user-images.githubusercontent.com/3118014/101991992-1fee3380-3c65-11eb-85d9-68c32aea1747.png)

**Comment by rnyoakum** [comment]
> Or maybe it could be like the (Orbit) missions - on the turn that you assemble the mission, if EVA suits aren't available, it would drop the EVA part of the mission.
>
> It might also be good to add a warning at the end of your turn that you have no EVA suits program, if you have a Manned Orbital* or any EVA mission planned for next turn. I'll add that to the End-of-Turn Warnings issue.

The game won't let you intentionally EVA if you don't have suits. Whether it stops you in the VAB/VIB as presently implemented or in Future Missions is a matter of choice. The problem is that a few missions will include an unexpected EVA in certain failure cases, for non-EVA missions. Unless you're proposing a (new) EVA suit requirement for all manned LM missions where an emergency EVA might be triggered?

**Comment by peyre** [comment]
Yeah, that sounds right...hum.
